### PR TITLE
Initial implementation of TACmi binding.

### PR DESCRIPTION
**Pull request by twendt**
This is an initial release of a binding for the CMI from Technische Alternative.

### CODE REVIEW

**Comment by teichsta** [comment]
Hi @twendt, thanks for this contribution! Since we are currently preparing the 1.7 release, i am pretty busy with code reviews right now. Also we announced a feature freeze for 1.7 was was due on April 19th.   I am more than happy to review the code to be introduced in 1.8 Snapshot (after May 25th). Best, Thomas E.-E.

**Comment by teichsta** [comment]
… sorry for letting you waiting so long. Will start reviewing your binding now.

File: bundles/binding/org.openhab.binding.tacmi/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -0,0 +1,28 @@
+Manifest-Version: 1.0
+Private-Package: org.openhab.binding.tacmi.internal
+Ignore-Package: org.openhab.binding.tacmi.internal
+Bundle-License: http://www.eclipse.org/legal/epl-v10.html
+Bundle-Name: openHAB TACmi Binding
+Bundle-SymbolicName: org.openhab.binding.tacmi
+Bundle-Vendor: openHAB.org
+Bundle-Version: 1.7.0.qualifier
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/META-INF/MANIFEST.MF]: bump version to 1.8

File: bundles/binding/org.openhab.binding.tacmi/pom.xml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+	<parent>
+		<groupId>org.openhab.bundles</groupId>
+		<artifactId>binding</artifactId>
+		<version>1.7.0-SNAPSHOT</version>
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/pom.xml]: bump version to 1.8

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/TACmiBindingProvider.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.tacmi;
+
+import org.openhab.binding.tacmi.internal.TACmiMeasureType;
+import org.openhab.core.binding.BindingProvider;
+
+/**
+ * @author Timo Wendt
+ * @since 1.7.0
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/TACmiBindingProvider.java]: 1.8

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+import org.osgi.framework.BundleContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implement this class if you are going create an actively polling service like
+ * querying a Website/Device.
+ * 
+ * @author Timo Wendt
+ * @since 1.7.0
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java]: 1.8

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+	 *            ConfigAdmin service
+	 */
+	public void activate(final BundleContext bundleContext,
+			final Map<String, Object> configuration) {
+		this.bundleContext = bundleContext;
+
+		// the configuration is guaranteed not to be null, because the component
+		// definition has the
+		// configuration-policy set to require. If set to 'optional' then the
+		// configuration may be null
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java]: remove generic comment

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+
+		// to override the default refresh interval one has to add a
+		// parameter to openhab.cfg like <bindingName>:refresh=<intervalInMs>
+		String refreshIntervalString = (String) configuration.get("refresh");
+		if (StringUtils.isNotBlank(refreshIntervalString)) {
+			refreshInterval = Long.parseLong(refreshIntervalString);
+		}
+
+		// set cmiAddress from configuration
+		cmiAddress = (String) configuration.get("cmiAddress");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java]: i would assume the binding won't work without the cmiAddress, right? If yes, please check the validity of the configuration and write some WARN logs if the configuration is invalid.

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+
+		// read further config parameters here ...
+		try {
+			clientSocket = new DatagramSocket(cmiPort);
+		} catch (SocketException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java]: replace with ```logger.warn()```

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+	/**
+	 * Called by the SCR when the configuration of a binding has been changed
+	 * through the ConfigAdmin service.
+	 * 
+	 * @param configuration
+	 *            Updated configuration properties
+	 */
+	public void modified(final Map<String, Object> configuration) {
+		// update the internal configuration accordingly
+	}
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java]: remove method (and also it's declaration in ```binding.xml```

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+										OnOffType.OFF);
+							}
+						} else {
+							logger.debug(
+									"CAN Node {} does not match or item not digital",
+									message.canNode);
+						}
+					}
+				}
+			}
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java]: there seem to be quite a lot code duplication in ```if``` and ```else if``` when moving the loops outside the if-else the duplication could be reduced (if not avoided)

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+		} catch (Exception e) {
+			logger.error("Error in execute: {}, Message: {}", e.getClass()
+					.getName(), e.getMessage());
+			e.printStackTrace();
+		}
+		logger.debug("TACmi execute() finished");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java]: decrease to TRACE

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+									"CAN Node {} does not match or item not digital",
+									message.canNode);
+						}
+					}
+				}
+			}
+		} catch (Exception e) {
+			logger.error("Error in execute: {}, Message: {}", e.getClass()
+					.getName(), e.getMessage());
+			e.printStackTrace();
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java]: decrease to WARN and remove ```printStackTrace```

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+	protected String getName() {
+		return "TACmi Refresh Service";
+	}
+
+	/**
+	 * @{inheritDoc
+	 */
+	@Override
+	protected void execute() {
+		logger.debug("execute() method is called!");
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java]: decrease to TRACE

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+					byte[] messageBytes = message.getRaw();
+					InetAddress IPAddress = InetAddress
+							.getByName(this.cmiAddress);
+					DatagramPacket packet = new DatagramPacket(messageBytes,
+							messageBytes.length, IPAddress,
+							TACmiBinding.cmiPort);
+					clientSocket.send(packet);
+				} catch (Exception e) {
+					logger.debug("Error sending message: {}, {}", e.getClass()
+							.getName(), e.getMessage());
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java]: increase to WARN

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+	}
+
+	/**
+	 * @{inheritDoc
+	 */
+	@Override
+	protected void internalReceiveUpdate(String itemName, State newState) {
+		logger.debug("internalReceiveUpdate({},{}) is called!", itemName,
+				newState);
+	}
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiBinding.java]: remove method

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiGenericBindingProvider.java [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+ * <li>
+ * <code>{ tacmi="&lt;can_node&gt;#d#&lt;port&gt;#Measuretype" }</code></li>
+ * <ul>
+ * <li>
+ * <code>{ tacmi="50#a#1#Temperature" }</code></li>
+ * </ul>
+ * </li> </ul>
+ * 
+ * @author Timo Wendt
+ * @since 1.7.0
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiGenericBindingProvider.java]: 1.8

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiGenericBindingProvider.java [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+			config.portNumber = Integer.parseInt(configParts[2]);
+			config.measureType = TACmiMeasureType.fromString(configParts[3]);
+			break;
+		default:
+			throw new BindingConfigParseException(
+					"A TACmi binding configuration must consist of three or four parts - please verify your *.items file");
+		}
+		config.configurationString = bindingConfig;
+
+		logger.debug("Adding binding: {}", config);
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiGenericBindingProvider.java]: > "Adding Binding configuration"

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiMeasureType.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.tacmi.internal;
+
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * @author Timo Wendt
+ * @since 1.7.0
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiMeasureType.java]: 1.8

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiMeasureType.java [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+ * 
+ *        This enum holds all the different measures and states available to be
+ *        retrieved by the TACmi binding Not the complete list yet.
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/TACmiMeasureType.java]: move above author tag

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/message/AnalogMessage.java [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+ * @since 1.7.0
+ */
+
+/**
+ * This class can be used to decode the analog values received in a message and
+ * also to create a new AnalogMessage used to send analog values to an analog
+ * CAN Input port. Creation of new message is not implemented so far.
+ * 
+ * @author Timo Wendt
+ * @since 1.7.0
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/message/AnalogMessage.java]: 1.8

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/message/AnalogValue.java [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+import org.openhab.binding.tacmi.internal.TACmiMeasureType;
+import org.openhab.core.library.types.DecimalType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class handles analog values as used in the analog message.
+ * 
+ * @author Timo Wendt
+ * @since 1.7.0
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/message/AnalogValue.java]: 1.8

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/message/DigitalMessage.java [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+
+import org.slf4j.Logger;
+
+/**
+ * This class can be used to decode the digital values received in a messag and
+ * also to create a new DigitalMessage used to send ON/OFF to an digital CAN
+ * Input port
+ * 
+ * @author Timo Wendt
+ * @since 1.7.0
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/message/DigitalMessage.java]: 1.8

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/message/Message.java [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+import java.nio.ByteOrder;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Base message class handling generic functions.
+ * 
+ * @author Timo Wendt
+ * @since 1.7.0
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/message/Message.java]: 1.8

File: bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/message/MessageType.java [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.tacmi.internal.message;
+
+/**
+* This enumeration represents the different message types provided by the MAX!Cube protocol. 
+* 
+* @author Timo Wendt
+* @since 1.7.0
```
Comment (teichsta) [inline bundles/binding/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/message/MessageType.java]: 1.8

**Comment by teichsta** [comment]
The binding looks quite good to me already! Please find some minor review comments inline and some general below:

* add wiki page (merely the same content like your README.md
* add binding to ```distribution/pom.xml```
* add possible configuration parameter to ```openhab_default.cfg```

Please finally squash your commits into one, once you've incorporated my feedback.

Thanks again, Thomas E.-E.

**Comment by twendt** [comment]
All the feedback has been incorporated now. I also created the Wiki page.

Thanks, Timo

**Comment by teichsta** [comment]
great, thanks a lot!

**Comment by teichsta** [comment]
> I also created the Wiki page.

where can in find it?

**Comment by twendt** [comment]
https://github.com/openhab/openhab/wiki/TA-Cmi-Binding

**Comment by Wolfgang1966** [comment]
Hello Timo,

I am extending the TA-CMI binding by more measure types currently. As I didn't find the measure type from the config used later I wonder whether it is really needed there or whether we lose something if it is not present in the config at all.

If possble, please contact me directly: tacmi [at] klimt [dot] de.

Thanks

Wolfgagn

**Comment by Wolfgang1966** [comment]
Forget my question. The measure type is needed for writing values.

Ok, then I know how to proceed...

**Pull request merged**